### PR TITLE
replace use of Visitor in initializerSemantic()

### PR DESCRIPTION
**Pull request by WalterBright**
Replaces with a switch statement. Should be faster, as it's static dispatch. It's less code, too!

### CODE REVIEW

**Comment by rainers** [comment]
> Should be faster, as it's static dispatch.

A switch with a number of cases is usually translated into a jump table which is just the same as virtual dispatch, but with the additional cost of having to boundscheck the table index.

**Comment by WalterBright** [comment]
> A switch with a number of cases is usually translated into a jump table which is just the same as virtual dispatch, but with the additional cost of having to boundscheck the table index.

Whether it's done that way or not depends on the pattern of the case values, and the compiler optimizer's choices. The bounds check is taken care of with branch prediction. The optimizer can also inline the functions, and optimize the aggregate.

The visitor pattern relies on double virtual dispatch.

@ibuclaw ran some benchmarks showing substantial improvement with switch https://github.com/dlang/dmd/pull/8659#issuecomment-425644075

**Comment by rainers** [comment]
> @ibuclaw ran some benchmarks showing substantial improvement with switch [#8659 (comment)](https://github.com/dlang/dmd/pull/8659#issuecomment-425644075)

I read his conclusion as "Time-wise, there's almost nothing that's different between before and after." The percentages just shifted from one place to another in the profiler results.

> The visitor pattern relies on double virtual dispatch.

I agree, double dispatch can be worse. The visitor pattern also adds overhead that can be eliminated if special cased.

**Comment by WalterBright** [comment]
I look at things a little differently. A regular rivet on an airplane does not slow it down perceptibly. But thousands do, which is why airplanes all switched to more expensive flush rivets around WW2. It's the same reason for the switch to titanium bolts instead a steel. Saving a gram here and there adds up to a major fuel savings.

I'm willing to spend a little performance cost if the advantages outweigh it. The only advantage I see here for Visitor is it doesn't need a discriminant, which seems a pretty small one.

File: src/dmd/initsem.d [hunk 1]
```diff
@@ -122,42 +122,18 @@ extern (C++) Expression initializerToExpression(Initializer init, Type t = null)
...
-    override void visit(VoidInitializer i)
+    Initializer visitVoid(VoidInitializer i)
     {
         i.type = t;
-        result = i;
+        return i;
     }
 
-    override void visit(ErrorInitializer i)
+    Initializer visitError(ErrorInitializer i)
```
[the remark this answers is not in the document]
Comment (WalterBright) [reply src/dmd/initsem.d]: Because nested functions can't be overloaded.

[the remark this answers is not in the document]
Comment (WalterBright) [reply src/dmd/initsem.d]:
```
void test() {
    void abc(int) { }
    void abc(long) { } // Error: declaration abc is already defined
}
```

[the remark this answers is not in the document]
Comment (jacob-carlborg) [reply src/dmd/initsem.d]: Ah, I missed that they weren't not defined in a class anymore.

**Comment by ibuclaw** [comment]
> I read his conclusion as "Time-wise, there's almost nothing that's different between before and after." The percentages just shifted from one place to another in the profiler results.

Actually I never said how long it took other than it was negligible. For phobos, both visitor and virtual implementations took around 6 seconds give or take 100ms to build phobos.

The switch implementation spends 2% less time in the function for the overall compilation. I forget how long it took, but that 2% translates to at least 120ms gain (around 5.8 seconds and less).

**Comment by wilzbach** [comment]
(Fwiw building DMD with LDC (and probably GDC too) leads to an easy 40-60% perf gain, but we still only ship DMD built with DMD.
So if anyone is interested in a better performance of DMD, that's the easiest and most effective fruit to pick...)

**Comment by n8sh** [comment]
>(Fwiw building DMD with LDC (and probably GDC too) leads to an easy 40-60% perf gain, but we still only ship DMD built with DMD.

Really? I thought GDC-built DMD was the standard because of the ~40% performance gain. Was that changed or was it never true?

**Comment by wilzbach** [comment]
You probably are:

https://github.com/dlang/installer/blob/235faa9a786858718f9d7937decbd0bc5b96a46c/create_dmd_release/build_all.d#L280

**Review by n8sh — APPROVED** [review]

**Review by RazvanN7 — APPROVED** [review]

**Pull request merged**